[R-package] installation via install.packages fails if /bin/sh is not bash (bashisms in configure)

## Description
An attempt to install lightgbm in an environment that has dash or some other posix compatible shell as /bin/sh will fail.

```
$ R -e "install.packages('lightgbm', version = '4.5.0', repos = 'https://cran.rstudio.com/')"
* installing *source* package 'lightgbm' ...
** package 'lightgbm' successfully unpacked and MD5 sums checked
** using staged installation
checking location of R... /usr/lib/R
checking whether MM_PREFETCH works... yes
./configure: line 1792: LGB_CPPFLAGS+= -DMM_PREFETCH=1: not found
checking whether MM_MALLOC works... yes
./configure: line 1827: LGB_CPPFLAGS+= -DMM_MALLOC=1: not found
configure: creating ./config.status                                                                    
config.status: creating src/Makevars     
** libs
using C++ compiler: 'g++ (Wolfi 14.2.0-r5) 14.2.0'                                 15:48:45 [1827/7522]
using C++17                                        
g++ -std=gnu++17 -I"/usr/include/R" -DNDEBUG -I./include -DEIGEN_MPL2_ONLY -DEIGEN_DONT_PARALLELIZE -DU
SE_SOCKET -DLGB_R_BUILD  -I/usr/local/include   -fopenmp -pthread -fpic  -g -O2   -c boosting/boosting.
cpp -o boosting/boosting.o                                                                             
In file included from ./include/LightGBM/config.h:21,                                                  
                 from ./include/LightGBM/boosting.h:8,                                                 
                 from boosting/boosting.cpp:5:
./include/LightGBM/utils/common.h:53:26: error: 'void* malloc(size_t)' was declared 'extern' and later 
'static' [-fpermissive]
   53 | #define _mm_malloc(a, b) malloc(a)
      |                          ^~~~~~
In file included from /usr/include/c++/14/cstdlib:79,
                 from /usr/include/c++/14/ext/string_conversions.h:43,
                 from /usr/include/c++/14/bits/basic_string.h:4154,
                 from /usr/include/c++/14/string:54,
                 from ./include/LightGBM/utils/json11.h:59,
                 from ./include/LightGBM/utils/common.h:8:
/usr/include/stdlib.h:672:14: note: previous declaration of 'void* malloc(size_t)'
  672 | extern void *malloc (size_t __size) __THROW __attribute_malloc__
      |              ^~~~~~
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/14/include/xmmintrin.h:34,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/14/include/emmintrin.h:31,
                 from /usr/lib/gcc/x86_64-pc-linux-gnu/14/include/pmmintrin.h:31,
                 from /usr/include/c++/14/x86_64-pc-linux-gnu/bits/opt_random.h:34,
                 from /usr/include/c++/14/random:49,
                 from ./include/LightGBM/utils/random.h:9,
                 from ./include/LightGBM/feature_group.h:11,
                 from ./include/LightGBM/dataset.h:10,
                 from ./include/LightGBM/objective_function.h:9,
                 from boosting/gbdt.h:9,
                 from boosting/dart.hpp:16,
                 from boosting/boosting.cpp:7:
/usr/lib/gcc/x86_64-pc-linux-gnu/14/include/mm_malloc.h: In function 'void* malloc(size_t)':
/usr/lib/gcc/x86_64-pc-linux-gnu/14/include/mm_malloc.h:41:7: error: '__alignment' was not declared in 
this scope
   41 |   if (__alignment == 1)
      |       ^~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/14/include/mm_malloc.h:43:7: error: '__alignment' was not declared in 
this scope
   43 |   if (__alignment == 2 || (sizeof (void *) == 8 && __alignment == 4))
      |       ^~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/14/include/mm_malloc.h:45:31: error: '__alignment' was not declared in
 this scope
   45 |   if (posix_memalign (&__ptr, __alignment, __size) == 0)
      |                               ^~~~~~~~~~~
/usr/lib/gcc/x86_64-pc-linux-gnu/14/include/mm_malloc.h: At global scope:
./include/LightGBM/utils/common.h:54:21: error: 'void free(void*)' was declared 'extern' and later 'sta
tic' [-fpermissive]
   54 | #define _mm_free(a) free(a)
      |                     ^~~~
/usr/include/stdlib.h:687:13: note: previous declaration of 'void free(void*)'
  687 | extern void free (void *__ptr) __THROW;
      |             ^~~~

make: *** [/usr/lib/R/etc/Makeconf:204: boosting/boosting.o] Error 1
ERROR: compilation failed for package 'lightgbm'
```

## Reproducible example

<!-- Minimal code that exhibits this behavior -->

## Environment info
Reproduced in wolfi environment, but will reproduce anywhere where /bin/sh is not bash.

LightGBM version or commit hash: 4.5.0

Command(s) you used to install LightGBM

```shell
R -e "install.packages('lightgbm', version = '4.5.0', repos = 'https://cran.rstudio.com/')"
```

## Additional Comments
Assignment with '+=' is not posix shell.
From the configure script:

```
#! /bin/sh
...
# line 1826
if test "${ac_mm_malloc}" = yes; then
   LGB_CPPFLAGS+=" -DMM_MALLOC=1"
fi
```

--- a/R-package/configure
+++ b/R-package/configure
@@ -1789,7 +1789,7 @@ ${CXX} ${CPPFLAGS} ${CXXFLAGS} -o conftest conftest.cpp 2>/dev/null && ./conftes
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: ${ac_mmprefetch}" >&5
 printf "%s\n" "${ac_mmprefetch}" >&6; }
 if test "${ac_mmprefetch}" = yes; then
-    LGB_CPPFLAGS+=" -DMM_PREFETCH=1"
+    LGB_CPPFLAGS="${LGB_CPPFLAGS} -DMM_PREFETCH=1"
 fi
 
 ############
@@ -1824,7 +1824,7 @@ ${CXX} ${CPPFLAGS} ${CXXFLAGS} -o conftest conftest.cpp 2>/dev/null && ./conftes
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: ${ac_mm_malloc}" >&5
 printf "%s\n" "${ac_mm_malloc}" >&6; }
 if test "${ac_mm_malloc}" = yes; then
-    LGB_CPPFLAGS+=" -DMM_MALLOC=1"
+    LGB_CPPFLAGS="${LGB_CPPFLAGS} -DMM_MALLOC=1"
 fi
 
 ##########
@@ -1850,11 +1850,11 @@ then
     # If Homebrew is found and libomp was installed with it, this code adds the necessary
     # flags for the compiler to find libomp headers and for the linker to find libomp.dylib.
     HOMEBREW_LIBOMP_PREFIX=""
-    if command -v brew &> /dev/null; then
+    if command -v brew >/dev/null 2>&1; then
         ac_brew_openmp=no
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking whether OpenMP was installed via Homebrew" >&5
 printf %s "checking whether OpenMP was installed via Homebrew... " >&6; }
-        brew --prefix libomp &>/dev/null && ac_brew_openmp=yes
+        brew --prefix libomp >/dev/null 2>&1 && ac_brew_openmp=yes
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: ${ac_brew_openmp}" >&5
 printf "%s\n" "${ac_brew_openmp}" >&6; }
         if test "${ac_brew_openmp}" = yes; then

--- a/R-package/configure.ac
+++ b/R-package/configure.ac
@@ -60,7 +60,7 @@ AC_LANG_CONFTEST(
 ${CXX} ${CPPFLAGS} ${CXXFLAGS} -o conftest conftest.cpp 2>/dev/null && ./conftest && ac_mmprefetch=yes
 AC_MSG_RESULT([${ac_mmprefetch}])
 if test "${ac_mmprefetch}" = yes; then
-    LGB_CPPFLAGS+=" -DMM_PREFETCH=1"
+    LGB_CPPFLAGS="${LGB_CPPFLAGS} -DMM_PREFETCH=1"
 fi
 
 ############
@@ -86,7 +86,7 @@ AC_LANG_CONFTEST(
 ${CXX} ${CPPFLAGS} ${CXXFLAGS} -o conftest conftest.cpp 2>/dev/null && ./conftest && ac_mm_malloc=yes
 AC_MSG_RESULT([${ac_mm_malloc}])
 if test "${ac_mm_malloc}" = yes; then
-    LGB_CPPFLAGS+=" -DMM_MALLOC=1"
+    LGB_CPPFLAGS="${LGB_CPPFLAGS} -DMM_MALLOC=1"
 fi
 
 ##########
@@ -112,10 +112,10 @@ then
     # If Homebrew is found and libomp was installed with it, this code adds the necessary
     # flags for the compiler to find libomp headers and for the linker to find libomp.dylib.
     HOMEBREW_LIBOMP_PREFIX=""
-    if command -v brew &> /dev/null; then
+    if command -v brew >/dev/null 2>&1; then
         ac_brew_openmp=no
         AC_MSG_CHECKING([whether OpenMP was installed via Homebrew])
-        brew --prefix libomp &>/dev/null && ac_brew_openmp=yes
+        brew --prefix libomp >/dev/null 2>&1 && ac_brew_openmp=yes
         AC_MSG_RESULT([${ac_brew_openmp}])
         if test "${ac_brew_openmp}" = yes; then
             HOMEBREW_LIBOMP_PREFIX=`brew --prefix libomp`

